### PR TITLE
frontend: cluster: hooks: Make useList multi cluster aware

### PR DESCRIPTION
**Pull request by illume**
Adds these keyed by cluster name to the return of useList():
- `clusterResults`, valued by result of useList on a single one.
- `clusterErrors`, valued by the error if any, otherwise null.

Adds the cluster field for the results from each cluster. So that the tables can display the cluster if needed.

The new useList is backwards compatible and works with one cluster too.

People can now also do this to get the results for cluster groups:
```ts
{ clusterErrors, clusterResults } = useList()
const error = clusterErrors[cluster]
clusterResults[cluster].items[0].cluster // this cluster field is now in each row
```

Another example of usage:
```ts
const { 
  items, // in multiple clusters mode items has .cluster field.
  error, // in single cluster mode this is a single error or null. In multi cluster it's the first error.
  clusterErrors // this is null if no errors, or a Record keyed by cluster name, with a value of the error.
} = resourceClass.useList();
``
```

### Testing

Note: for single clusters it uses the same function as before.

- [ ] in single mode the cluster list views work
- [ ] in multi mode the cluster list views show results from multiple clusters and there is a 'cluster' column in the results.

### CODE REVIEW

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (joaquimrocha) [inline frontend/src/lib/k8s/api/v2/hooks.ts:306]: I wonder if we will have problems with this given we are not using the hook in the function body. But we do use some like this already anyway...

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (joaquimrocha) [inline frontend/src/lib/k8s/api/v2/hooks.ts:355]: I think this is only for the return to work also as a list. Since there's no backward compat need for this function, we can remove this.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (joaquimrocha) [inline frontend/src/lib/k8s/api/v2/hooks.ts:290]: I had actually thought that we could adapt the useKubeObjectList and pass a new `clusters` arg. But then we'd have to use `useQueries` if we want to maintain a cache of each cluster's results independently and maybe that adds complexity. So maybe we can use this new function.
I'd like @sniok 's opinion here.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (joaquimrocha) [inline frontend/src/lib/k8s/api/v2/hooks.ts:65]: Maybe we could extend the previous interface, rather than adding more members to this one? Since it's a new function, having a new interface doesn't seem inadequate.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:355]: plugins may need backward compatibility

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: I wrapped it in useMemo to ensure consistent hook order and such when clusters change.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: I removed useKubeObjectLists, and now just useKubeObjectList is exported. The multi cluster version is internal.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:65]: I removed useKubeObjectLists, and now just useKubeObjectList is exported. The multi cluster version is internal.

**Comment by illume** [comment]
I removed useKubeObjectLists, and now just useKubeObjectList is exported. The multi cluster version is internal.

The tests are now passing in CI.

**Comment by illume** [comment]
todo: There's a better way to add the cluster field.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts:61]: is clusterResults ever used? I'd think that items list with all the responses is enough? 

this is a bit hard to review without seeing other changes

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts:351]: Having hooks in a loop is a bit sketchy (although we already have loops like that in just a couple of places where we also ensure that the parent component will remount) the useList is used pretty much everywhere. Ideally we'd use useQueries to fetch multiple queries at once but I think this is okay for now

Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:351]: When clusters change top level components remount. Which means the order is fixed.

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:61]: > this is a bit hard to review without seeing other changes to the ListViews of resources

You can see clusterErrors used here in this PR from the other day:
https://github.com/headlamp-k8s/headlamp/pull/2436

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (joaquimrocha) [inline frontend/src/lib/k8s/api/v2/hooks.ts:246]: Why is the cluster not set in the constructor? Is this the reason you added a cluster setter? I am asking because it feels like the cluster should be a readonly in that class.

**Review by joaquimrocha — APPROVED** [review]
It LGTM even if we should keep in mind to change the logic to use `useQueries` later on.
I left a comment about the cluster setter, but if there's a good reason to keep that, please feel free to move on.

**Review by sniok — APPROVED** [review]

File: frontend/src/lib/k8s/api/v2/hooks.ts
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:246]: Yeah... I had to do this because the typing is broken for that code unfortunately. It doesn't know that there is a constructor that takes an extra argument and complains.

I guess we can look back at this once the other typing fixes are merged.

**Pull request merged**
